R/travis-auth.R

```diff
@@ -1,75 +1,92 @@
+  cli::cli_alert_info(
+    "Waiting for API token to appear on the clipboard."
+  )
+  Sys.sleep(3)
 
-travis_token_ <- function(repo = NULL) {
-  token <- auth_travis()
-  if (!is.null(repo)) {
-    if (!travis_has_repo(repo, token)) {
-      travis_sync(token = token)
-      if (!travis_has_repo(repo, token)) {
-        review_travis_app_permission(repo)
-      }
+  repeat {
+    token <- readline("Please paste the API token to the console.\n")
+    if (is_token(token)) break
+    Sys.sleep(0.1)
+  }
+  cli::cli_alert("Detected API token. Clearing clipboard.")
+  requireNamespace("clipr", quietly = TRUE)
+  tryCatch(
+    clipr::write_clip(""),
+    error = function(e) {
+      warning("Error clearing clipboard: ", conditionMessage(e))
     }
+  )
+  dir.create("~/.travis", showWarnings = FALSE)
+
+  # if there is already a file with the API key of a different endpoint,
+  # we need to append only
+  has_conf <- file.exists("~/.travis/config.yml")
+  if (has_conf) {
+    cli::cli_alert_warning(
+      "Existing API token detected in {.file ~/.travis/config.yml}.
+      Appending new API token ({endpoint}).", wrap = TRUE
+    )
+
+    endpoint_line <- which(grepl(sprintf("endpoints", endpoint), readLines("~/.travis/config.yml")))
+    yml <- readLines("~/.travis/config.yml")
+    yml[endpoint_line] <- sprintf(
+      "endpoints:\n  https://api.travis-ci%s/:\n    access_token: %s",
+      endpoint, token
+    )
+    cli::cli_alert("Appending Travis CI API token for endpoint '{endpoint}' to
+                   {.file ~/.travis/config.yml}.", wrap = TRUE)
+    writeLines(yml, "~/.travis/config.yml")
+  } else {
+    cli::cli_alert("Storing Travis CI API token for endpoint '{endpoint}' in
+                  {.file ~/.travis/config.yml}.", wrap = TRUE)
+    writeLines(sprintf(
+      "endpoints:\n  https://api.travis-ci%s/:\n    access_token: %s",
+      endpoint, token
+    ), con = "~/.travis/config.yml")
   }
-  token
 }
 
-review_travis_app_permission <- function(repo) {
-  url_stop(
-    "You may need to retry in a few seconds. ",
-    "If your repo ", repo, " belongs to an organization, you may need to allow Travis access to that organization",
-    url = "https://github.com/settings/connections/applications/f244293c729d5066cf27"
-  )
+is_token <- function(token) {
+  grepl("^[-a-zA-Z0-9_]{18}", token)
 }
 
-#' @description
-#' `travis_token()` authenticates and checks if the repository is known to
-#' Travis CI. If not, a GitHub sync via [travis_sync()] is performed.
-#'
-#' @param repo `[string]`\cr
-#'   The GitHub repo slug in the format "<user|org>/<repo>", if `NULL` no sync
-#'   is performed.
-#'
-#' @export
-#' @rdname auth_travis
-travis_token <- memoise::memoise(travis_token_)
+# travis <- function(endpoint = "") {
+#   sprintf("https://api.travis-ci%s", endpoint)
+# }
+
+read_token <- function(endpoint) {
+  yml <- readLines("~/.travis/config.yml")
+  endpoint_line <- which(grepl(sprintf("api.travis.ci%s", endpoint), yml))
```

Comment:
🙈 🙉 🙊

Reply:
I know, does not look elegant but works and seems to be quite robust?